src/PowerShellEditorServices/Language/TokenOperations.cs

```diff
@@ -0,0 +1,271 @@
+        /// </summary>
+        static private FoldingReference CreateFoldingReference(
+            Token startToken,
+            Token endToken,
+            string matchKind)
+        {
+            if (endToken.Extent.EndLineNumber == startToken.Extent.StartLineNumber) { return null; }
+            // Extents are base 1, but LSP is base 0, so minus 1 off all lines and character positions
+            return new FoldingReference(
+                startToken.Extent.StartLineNumber - 1,
+                startToken.Extent.StartColumnNumber - 1,
+                endToken.Extent.EndLineNumber - 1,
+                endToken.Extent.EndColumnNumber - 1,
+                matchKind
+            );
+        }
+
+        /// <summary>
+        /// Creates an instance of a FoldingReference object from a start token and an end line
+        /// Returns null if the line range is invalid
+        /// </summary>
+        static private FoldingReference CreateFoldingReference(
+            Token startToken,
+            int endLine,
+            string matchKind)
+        {
+            if (endLine == (startToken.Extent.StartLineNumber - 1)) { return null; }
+            return new FoldingReference(
+                startToken.Extent.StartLineNumber - 1,   // Extents are base 1, but LSP is base 0
+                startToken.Extent.StartColumnNumber - 1, // Extents are base 1, but LSP is base 0
+                endLine,
+                0,
+                matchKind
+            );
+        }
+
+        /// <summary>
+        /// Given a array tokens finds matching regions which start and end with a different TokenKind
+        /// </summary>
+        static private List<FoldingReference> MatchTokenElements(
+            Token[] tokens,
+            TokenKind startTokenKind,
+            TokenKind endTokenKind,
+            string matchKind)
+        {
+            List<FoldingReference> result = new List<FoldingReference>();
+            Stack<Token> tokenStack = new Stack<Token>();
+            foreach (Token token in tokens)
+            {
+                if (token.Kind == startTokenKind) {
+                    tokenStack.Push(token);
+                }
+                if ((tokenStack.Count > 0) & (token.Kind == endTokenKind)) {
+                    result.Add(CreateFoldingReference(tokenStack.Pop(), token, matchKind));
+                }
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Given a array tokens finds a specific token
+        /// </summary>
+        static private List<FoldingReference> MatchTokenElement(
+            Token[] tokens,
+            TokenKind tokenKind,
+            string matchKind)
+        {
+            List<FoldingReference> result = new List<FoldingReference>();
+            foreach (Token token in tokens)
+            {
+                if ((token.Kind == tokenKind) & (token.Extent.StartLineNumber != token.Extent.EndLineNumber)) {
```

Comment:
I noticed that you're using the Boolean logical AND operator. Can you just use `&&` instead? Then if x is false, it wont go to y.

Reply:
```suggestion
                if ((token.Kind == tokenKind) && (token.Extent.StartLineNumber != token.Extent.EndLineNumber)) {
```

Reply:
Whoops...yeah that should be `&&`

Reply:
Fixed